src/blockMonitor/blockCache.ts

```diff
@@ -0,0 +1,230 @@
+import { ethers } from "ethers";
+import { ApplicationError } from "../dataEntities";
+import { IBlockStub, BlockStubChain } from "./blockStub";
+
+// adds item to map.get(key), but make sure it exists first
+function addItemToKeyedSet<T, U>(map: Map<T, Set<U>>, key: T, item: U) {
+    const set = map.get(key);
+    if (set === undefined) {
+        map.set(key, new Set([item]));
+    } else {
+        set.add(item);
+    }
+}
+
+// removes item from map.get(key), and also delete the resulting set if empty
+function removeItemFromKeyedSet<T, U>(map: Map<T, Set<U>>, key: T, item: U) {
+    const set = map.get(key);
+    if (set === undefined || set.delete(item) === false) {
+        throw new ApplicationError("Tried to remove item from a set that does not contain it.");
+    }
+    if (set.size === 0) {
+        map.delete(key);
+    }
+}
+
+/**
+ * Utility class to store and query info on full blocks up to a given maximum depth `maxDepth`, compared to the current
+ * maximum height ever seen.
+ * It prunes all the blocks at depth bigger than `maxDepth`, or with height smaller than the first block that was added.
+ * It does not allow to add blocks without adding their parent first, except if they are at depth `maxDepth`.
+ **/
+export class BlockCache {
+    public blockStubsByHash: Map<string, BlockStubChain> = new Map();
+
+    // store block hashes at a specific height (there could be more than one at some height because of forks)
+    public blockHashesByHeight: Map<number, Set<string>> = new Map();
+
+    // maps transaction hashes to blocks containing them
+    public blockHashesByTxHash: Map<string, Set<string>> = new Map();
+
+    // set of tx hashes per block hash, for fast lookup
+    public txHashesByBlockHash: Map<string, Set<string>> = new Map();
+
+    // Blocks at height smaller than which all blocks have already been pruned
+    private minStoredHeight = 0;
+
+    // After the first block is added, its height. Will not store blocks with smaller height
+    private initialHeight: number = -1;
+
+    // Height of the highest known block
+    private mMaxHeight = 0;
+    public get maxHeight() {
```

Comment:
Personally I would leave it there as harmless, as otherwise a user of the blockCache who was not the one adding blocks has no way of knowing about the current height of the cache.